### ISSUE TITLE
Use xarray (w/dask) in workflow

### ISSUE THREAD

**Issue by ggg121** [radical-collaboration/facts#25]
Many geoscientific datasets (such as the CMIP6 archive) are in netCDF format.  It's likely one would need to open multiple netCDF files within their workflow, potentially loading 10s to 100s of GB of data into system memory.

The `xarray` Python module addresses this potential problem, along with some quality-of-life improvements to the way these datasets are handled and analyzed.  The `xarray` module leverages the `dask` Python module to "chunk" the datasets into smaller pieces and perform operations on these smaller chunks.  No operations are performed until the values need to be calculated (lazy `xarray` vs. aggressive `numpy`).  This means, less system memory will be used to perform the same operations.  The `xarray` module is gaining popularity among geoscientists and it's very likely that a FACTS module may need it.  Which begs the question...

Would EnTK be able to handle a workflow that uses the `xarray` Python module (with `dask`)?

**Comment by ggg121** [comment]
I've tested this a bit (see "test_xarray" in "other" directory under the devel branch).  The workflow consists of two identical pipelines.  Each has a single stage with two tasks.  Each task is the same - run the script "test_xarray.py" using the 600+ MB netCDF file, which reads in the input netCDF file in chunks, analyzes the data, and produces a small netCDF file as output.  I've run this locally (2 cpus), and twice on a remote machine that uses slurm (once with 4 cpus and once with 36 cpus to ensure engagement with multiple nodes).  Each test returned the expected output files without any apparent problems.

Are there any additional (potentially problematic) situations I should try to emulate with these tests?  If not, I'm sufficiently convinced that, for my application, there's no negative interaction between xarray w/dask and entk.

**Comment by mturilli** [comment]
Thanks for doing this Greg. Would you mind sharing the profiles of your largest test? I am assuming you enabled profiling?

**Comment by ggg121** [comment]
I've performed 4 experiments, all of which were run on Amarel:

1. (018432.0002) Request 16 CPUs from application manager, assign 4 processes per task through Task.cpu_reqs, enable profiling and debugging.
2. (018432.0003) Request 16 CPUs from application manager, assign 1 processes per task through Task.cpu_reqs, enable profiling and debugging.
3. (018435.0000) Request 16 CPUs from application manager, do not assign anything to Task.cpu_reqs, enable profiling and debugging.
4. (018435.0001) Request 16 CPUs from application manager, do not assign anything to Task.cpu_reqs, do not enable profiling and debugging.

All tasks fail in experiment (1).  All tasks complete in experiments (2) - (4) in that the expected output is downloaded to the local machine; however, something hangs until the requested walltime expires.  I've uploaded the session information, including the runtime verbose log files, from my local machine to Amarel.  I'll point @Abdullah-Ghani to their location on Amarel for further investigation.
